Add Aborted conditionType for policy enactments

rhrazdil opened the pull request:
Signed-off-by: Radim Hrazdil <rhrazdil@redhat.com>



**Is this a BUG FIX or a FEATURE ?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
/kind enhancement

**What this PR does / why we need it**:
When rolling out nncp sequentially, if a node fails to apply configuration
other nodes do not attempt to apply it and set their enactment condition to
Aborted.

Example:
We apply sequential `policy` in a 3 node cluster.
Assume the fist node that starts applying configuration is node01, then the result will look like:
```
$ kubectl get nnce
NAME            STATUS
node01.policy   FailedToConfigure
node02.policy   ConfigurationAborted
node03.policy   ConfigurationAborted
```
In other words, node01 has started applying and failed. Node02 and node03 noted that there is a node failing to apply `policy`
and instead of attempting to apply it too set their enactment status to `ConfigurationAborted`.

**Special notes for your reviewer**:

**Release note**:


```release-note
Added new conditionType Aborted used in sequential configuration rollout
```

qinqon commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
```
Comment not needed since the error explain it.

qinqon commented on pkg/policyconditions/conditions.go:
```diff
@@ -152,8 +152,11 @@ func Update(cli client.Client, policyKey types.NamespacedName) error {
+			} else if enactmentsCount.Failed() > 0 || enactmentsCount.Aborted() > 0 {
 				message := fmt.Sprintf("%d/%d nodes failed to configure", enactmentsCount.Failed(), enactmentsCount.Matching())
+				if enactmentsCount.Aborted() > 0 {
+					message += fmt.Sprintf(", %d nodes aborted configuration", enactmentsCount.Aborted())
```
Let's add also the number of matching

rhrazdil replied:
it is still there, I concatenate it to the original `message`

qinqon reviewed (changes requested):
first pass: Skipping review of tests
some nits and a question
but looks pretty good.

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -161,19 +183,6 @@ func (r *NodeNetworkConfigurationPolicyReconciler) setNodeRunningUpdate(policyKe
-	policy *nmstatev1beta1.NodeNetworkConfigurationPolicy,
-	ec *enactmentconditions.EnactmentConditions,
-) error {
-	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
-		// a node has been running the update for too long
-		err := fmt.Errorf("A node has been configuring for too long, aborting")
-		ec.NotifyFailedToConfigure(err)
-		return err
```
I only moved it to another function called before `claimNodeRunningUpdate`. I didn't like that function that is named `claimSomething` does a lot of other things apart of claimingSomething.

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
 	}
 	if policy.Status.NodeRunningUpdate != "" {
-		return fmt.Errorf("another node is working on configuration")
+		return apierrors.NewConflict(schema.GroupResource{Resource: "nodenetworkconfigurationpolicies"}, policy.Name, fmt.Errorf("Another node is working on configuration"))
```
Yes indeed

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+		err := fmt.Errorf("current node has failed to apply policy configuration")
+		return err
+	}
+	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
```
That's a good point. It looks like we don't need the timestamp at all then

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+	}
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
```
Done

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
+		err := fmt.Errorf("current node has failed to apply policy configuration")
+		return err
```
Done

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+	if policy.Status.NodeRunningUpdate != "" && metav1.Now().Sub(policy.Status.NodeUpdateStart.Time) > nodeConfigurationTimeout {
+		// a node has been running the update for too long
+		err := fmt.Errorf("current node has been configuring for too long")
+		return err
```
Done

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -143,14 +144,35 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+	enactmentsCount := enactmentconditions.Count(enactments, policy.Generation)
+	if enactmentsCount.Failed() > 0 {
+		// a node has failed to apply configuration
+		err := fmt.Errorf("current node has failed to apply policy configuration")
```
Right, done

qinqon commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -55,7 +56,6 @@ import (
 
 var (
 	nodeName                                string
-	nodeConfigurationTimeout                = 5 * time.Minute
```
We don't need this ?

rhrazdil replied:
Since we don't need to check how long a node is applying the policy, I think we can remove it, it's not being used except for setting it

qinqon replied:
You mean that we already have a timeout mechanism when we run `nmstatectl set` so we don't need any of this right ?

rhrazdil replied:
Yes, I believe it's enough to rely on the timeout of `ApplyDesiredState` and only check if there is a node with Failed enactment

qinqon reviewed (changes requested):
We don't need the configuration timeout anymore ? 

Also looks like docs/assets/images/logo/fullcolor.png has being overwritten.

rhrazdil commented:
> Also looks like docs/assets/images/logo/fullcolor.png has being overwritten.

you're right, thanks. Removed this change

rhrazdil commented:
in the last push I have
- removed the timestamp of claiming nodeRunningUpdate
- fixed the test that failed in parallel lane, it wasn't expecting the aborted state to be there

qinqon commented on pkg/enactmentstatus/conditions/counter_test.go:
```diff
@@ -154,6 +162,35 @@ var _ = Describe("Enactment condition counter", func() {
+				aborted:     CountByConditionStatus{t: 1, f: 1, u: 0},
+			},
+		}),
+		Entry("e(Matching, Progressing), e(Matching, Aborted)", EnactmentCounterCase{
```
Cannot be progressing + aborted I think we can remove this test.

rhrazdil replied:
Done

qinqon commented on test/e2e/handler/nnce_conditions_test.go:
```diff
@@ -110,11 +141,33 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
+			for i, _ := range nodes {
+				node := nodes[i]
+				go func() {
+					defer wg.Done()
```
we are missing the `defer GinkgoRecover`

rhrazdil replied:
Added

qinqon commented on test/e2e/handler/nnce_conditions_test.go:
```diff
@@ -128,23 +181,36 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
 			waitForDegradedTestPolicy()
 
-			By("Check that the enactment stays in failing state")
+			for _, node := range nodes {
+				By(fmt.Sprintf("Check %s failing or aborted state is reached", node))
+				enactmentConditionsStatusEventually(node).Should(
+					SatisfyAny(
+						ConsistOf(failingEnactmentConditions...),
+						ConsistOf(abortedEnactmentConditions...),
+					), "should eventually reach failing of aborted conditions at enactments")
+			}
+
+			By("Check that the enactment stays in failing or aborted state")
```
This will pass is something is broken and we have two failures and one aborted at sequencial rollout we will have just one failed enactment.

rhrazdil replied:
Addressed by adding a function that counts failing and aborted enactments and expects 1 failing and len(nodes)-1 aborted enactments

qinqon commented on test/e2e/handler/nnce_conditions_test.go:
```diff
@@ -128,23 +181,36 @@ var _ = Describe("[rfe_id:3503][crit:medium][vendor:cnv-qe@redhat.com][level:com
 					defer wg.Done()
 					By(fmt.Sprintf("Check %s failing state is kept", node))
-					enactmentConditionsStatusConsistently(node).Should(ConsistOf(failingEnactmentConditions...), "should consistently keep failing conditions at enactments")
+					enactmentConditionsStatusConsistently(node).Should(
+						SatisfyAny(
+							ConsistOf(failingEnactmentConditions...),
+							ConsistOf(abortedEnactmentConditions...),
+						), "should consistently keep failing or aborted conditions at enactments")
```
ditto

rhrazdil replied:
I think in this case, we can keep the or, because should any of those enactment change, it would need to be at least in `matching` or `notMatching`.
Afterwards I again check the numbers of aborted and failing

qinqon reviewed (changes requested):
Looking even better, I see code shrinking :-), e2e test reviewed too.

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -166,48 +165,45 @@ func (r *NodeNetworkConfigurationPolicyReconciler) getEnactmentCount(policy *nms
+		return fmt.Errorf("some node/s has failed to apply policy configuration")
+	}
+	return nil
+}
```
Done

rhrazdil commented on controllers/nodenetworkconfigurationpolicy_controller.go:
```diff
@@ -263,9 +259,15 @@ func (r *NodeNetworkConfigurationPolicyReconciler) Reconcile(request ctrl.Reques
+			enactmentConditions.NotifyAborted(err)
+			return ctrl.Result{}, nil
+		}
+		err = r.claimNodeRunningUpdate(instance)
```
Makes sense, thanks. Done

rhrazdil commented:
/test pull-kubernetes-nmstate-e2e-handler-k8s

qinqon commented:
/lgtm
/approve
